back/tests/WebSpec.hs

```diff
@@ -25,23 +26,92 @@ import Control.Monad.Catch
 
 -- Testing
 import Selenium
+import qualified Api
 import qualified Test.WebDriver.Common.Keys as Key
 
 -- Site
 import qualified Guide.Main
-import Guide.Config (Config(..))
+import Guide.Logger
+import Guide.Config (Config(..), def)
+import Text.RE.TDFA.String
 
+-- Spec
+import qualified Test.Hspec as H
+import System.IO
 
 -----------------------------------------------------------------------------
 -- Tests
 -----------------------------------------------------------------------------
 
+logFile :: FilePath
+logFile = "/tmp/test_guide.log"
+
 tests :: IO ()
-tests = run $ do
-  mainPageTests
-  categoryTests
-  itemTests
-  markdownTests
+tests = do
+  run $ do
+    mainPageTests
+    categoryTests
+    itemTests
+    markdownTests
+    Api.tests
+  hspec logTest
+
+getLines :: Handle -> IO String
+getLines h = loop' []
+  where
+    loop' :: [String] -> IO String
+    loop' xs = do
+      eLine <- try $ hGetLine h
+      case eLine of
+        Left (_ :: SomeException) -> pure $ concat $ reverse xs
+        Right line                -> loop' (line:xs)
+
+logTest :: Spec
+logTest = H.describe "test of logger" $ do
+  logs <- H.runIO $ do
+    logFileHandle <- openFile logFile ReadWriteMode
+    logs <- getLines logFileHandle
+    hClose logFileHandle
+    writeFile logFile ""
+    pure logs
+  H.describe "Logging of init" $ do
+    H.it "spock is runned" $ [re|Spock is running on port|] `isIn` logs
+    H.it "api is runned" $ [re|API is running on port|] `isIn` logs
+  H.describe "Logging of api" $ do
+    H.describe "Categories" $ do
+      H.it "get categories request" $ [re|getCategories|] `isIn` logs
+      H.it "create ctegory request" $ [re|createCategory|] `isIn` logs
+      H.it "get category by id" $ [re|getCategory|] `isIn` logs
+      H.it "delete category by id request" $ [re|deleteCategory|] `isIn` logs
+      H.it "modify notes to category request" $ [re|setCategoryNotes|] `isIn` logs
+      H.it "modify info to category request" $ [re|setCategoryInfo|] `isIn` logs
+    H.describe "Item" $ do
```

Comment:
I feel it's not necessary to test logging for every single endpoint.

If somebody is careless enough that they see 20 endpoints with logging and still forget to add logging to the new endpoint they're writing, they will also sure as heck forget to add a test for it.

Reply:
So, either will we stop to add new tests or will we delete several to be not the whole?

Reply:
If they are testing only logging, we can delete all of them (well, leave one).